### DISCUSSION TITLE
Tool recommendation for generating code coverage for iOS projects?

### DISCUSSION THREAD

**Discussion by antranapp** [MobileNativeFoundation/discussions#138] [General]
Anyone has good experiences with a CLI tool for generating code coverage for iOS projects?

The native way to generate Code Coverage with xcodebuild can't ignore certain folders/code.

I'm aware of [Slather](https://github.com/SlatherOrg/slather) which supports many configuration possibilities.

But I'd like to ask folks out there if they have any alternative solution which:
- Can generate code coverage reports that are suitable to consume on CI (we want to show the percentage badge)
- Can generate human readable code hit/miss report so that we know where we missed a test
- Ideally, CLI
- has possibilities to configure ignorance so that we can focus on the code we care about.

(I know that Slather supports all of those features, just want to make sure that I don't miss any other tools that can do something similar).

**Comment by keith** [comment]
I haven't seen any out-of-the-box tools for this, but realistically to have enough control for what it sounds like you want, you should probably invoke [`llvm-cov`](https://llvm.org/docs/CommandGuide/llvm-cov.html) (which is likely what Xcode is doing under the hood) yourself. 

For example in our setup we do something like this:

```
xcrun llvm-cov \
  show \
  -format html \
  -instr-profile path/to/foo.profdata \
  -path-equivalence=.,"$PWD" \
  -output-dir=whatever \
  path/to/foo.xctest/binary \
  @"$sources_only_manifest"
```

If you're using vanilla xcodebuild you can find the profdata, and intermediate binary in derived data, and the `sources_only_manifest` in this example is a text file with a filepath on each line that is used to decide what files to limit the output to. This example outputs a html dump, but there are a bunch of other options depending on if you just want some human readable text based output, or json you can parse.

You can also generate `lcov` files which can be used with other [tools](https://github.com/linux-test-project/lcov) if you'd like. Here's another example https://github.com/bazelbuild/rules_apple/pull/1223
